### PR TITLE
Adding a get_random_community endpoint.

### LINKED ISSUE #4698: Get random community
### Requirements

- [X] Is this a feature request? For questions or discussions use https://lemmy.ml/c/lemmy_support
- [X] Did you check to see if this issue already exists?
- [X] Is this only a feature request? Do not put multiple feature requests in one issue.
- [X] Is this a backend issue? Use the [lemmy-ui](https://github.com/LemmyNet/lemmy-ui) repo for UI / frontend issues.
- [X] Do you agree to follow the rules in our [Code of Conduct](https://join-lemmy.org/docs/code_of_conduct.html)?

### Is your proposal related to a problem?

I would like a way to fetch a random community

### Describe the solution you'd like.

`GET /api/v3/community/random?type_=All`

Returns:

`GetCommunityResponse`

### Describe alternatives you've considered.

There is no good alternative. You could get a random page of communities, but there is no way to determine the total number of federated communities for a site (you can only determine # of local communities in site aggregates).

Also, page # is being phased out in favor of page cursor (which is good, but will make the above workaround no longer possible)

### Additional context

There is definite user interest for this feature with Voyager.

### PR DESCRIPTION
**Pull request by dessalines**
- Fixes #4698

### CODE REVIEW

File: crates/db_schema/src/impls/community.rs [hunk 1]
```diff
@@ -194,6 +193,20 @@ impl Community {
...
+    sql_function!(fn random() -> Text);
+    community::table
+      .filter(community::local)
+      .filter(not(community::deleted))
+      .filter(not(community::removed))
+      .order(random())
+      .limit(1)
+      .first::<Self>(conn)
+      .await
+      .optional()
```
Comment (Nutomic) [inline crates/db_schema/src/impls/community.rs:208]: Get rid of optional (see https://github.com/LemmyNet/lemmy/pull/5031)

Comment (dessalines) [reply crates/db_schema/src/impls/community.rs]: Done.

File: crates/db_schema/src/impls/community.rs [hunk 2]
```diff
@@ -193,6 +193,19 @@ impl Community {
       .await?;
     Ok(())
   }
+
+  pub async fn get_random_local_community(pool: &mut DbPool<'_>) -> Result<Self, Error> {
+    let conn = &mut get_conn(pool).await?;
+    sql_function!(fn random() -> Text);
+    community::table
+      .filter(community::local)
```
Comment (Nutomic) [inline crates/db_schema/src/impls/community.rs:201]: The issue doesnt mention that it should return only local communities, and thats not clear from the endpoint name either. Based on the issue it should take a type param (`GET /api/v3/community/random?type_=All`)

File: crates/db_schema/src/impls/community.rs [hunk 3]
```diff
@@ -193,6 +193,19 @@ impl Community {
       .await?;
     Ok(())
   }
+
+  pub async fn get_random_local_community(pool: &mut DbPool<'_>) -> Result<Self, Error> {
```
Comment (Nutomic) [inline crates/db_schema/src/impls/community.rs:197]: Its enough to select CommunityId.

File: crates/db_schema/src/impls/community.rs (hunk 2 above)
Comment (dessalines) [reply crates/db_schema/src/impls/community.rs]: K, added ListingType to the query.

File: crates/db_schema/src/impls/community.rs (hunk 3 above)
Comment (dessalines) [reply crates/db_schema/src/impls/community.rs]: Done.

File: crates/db_schema/src/impls/community.rs [hunk 4]
```diff
@@ -193,6 +194,33 @@ impl Community {
...
+      .filter(not(community::deleted))
+      .filter(not(community::removed))
+      .into_boxed();
+
+    if let Some(type_) = type_ {
+      query = match type_ {
+        ListingType::Local => query.filter(community::local),
+        _ => query,
+      };
+    };
```
Comment (Nutomic) [inline crates/db_schema/src/impls/community.rs:215]:
```suggestion
      query = match type_ {
        Some(ListingType::Local) => query.filter(community::local),
        _ => query,
      };
```

Comment (dessalines) [reply crates/db_schema/src/impls/community.rs]: K I added a more concise way.

**Review by Nutomic — APPROVED** [review]

**Pull request merged**